[쿨라임] 로또 1단계 - 자동 미션 제출합니다.

icyMojito opened the pull request:
감사합니다!

Commit:
docs : README 작성

Commit:
docs : README 수정

Commit:
feat : 로또 구입 금액을 입력받고 검증 후 예외처리하는 PurchasePrice 클래스 생성

Commit:
feat : 로또 구입 개수를 계산하는 메서드 추가, 출력하는 메서드 추가

Commit:
feat : 로또 한 줄을 생성하는 LottoGenerator 클래스 생성

Commit:
feat : 로또 한 줄의 정보를 담는 Lotto 클래스 생성

Commit:
feat : 로또 한 줄의 숫자를 출력하는 메서드를 뷰에 추가

Commit:
feat : 구입한 모든 로또를 관리하는 Lottos 클래스 생성

Commit:
feat : 구입한 모든 로또의 번호를 출력하는 printLottos() 생성

Commit:
docs : README 수정

당첨 번호와 보너스 번호에 대한 예외 추가 정리

Commit:
feat : 당첨 번호를 입력받아 검증하는 WinningNumbers 클래스 생성, 테스트코드 작성

Commit:
feat : 보너스 번호를 검증하는 BonusNumber 클래스 생성

Commit:
docs : README 수정

Commit:
feat : 당첨 결과를 계산하는 로직 추가

Lotto클래스에 로또 한 줄과 당첨 번호를 비교해서 비교 결과를 반환하는 메서드 추가
Lottos클래스에 로또 여러줄과 당첨 번호를 비교해서 전체 로또 결과를 반환하는 메서드 추가

Commit:
feat : 당첨 결과 계산을 위한 메서드 추가, 결과를 나타내는 MatchResult enum 생성

Commit:
test : 당첨 결과 반환 로직 테스트

Commit:
feat : 당첨 결과를 관리하고 수익률을 계산하는 LottoResult 클래스 생성

Commit:
test : 수익률에 관한 테스트 추가

Commit:
feat : 해당되는 당첨 결과가 없을 때 처리 로직 추가

Commit:
feat : 로또 당첨 결과와 수익률 출력 기능 추가

Commit:
refactor : 메서드 분리와 순서, 상수 정리

Commit:
test : 테스트 코드 상수명 변경과 필요없는 공백 삭제

Commit:
docs : README 수정

icyMojito committed:
Merge pull request #1 from hwanghe159/hwanghe159

Hwanghe159

kang-hyungu commented on src/main/java/lotto/domain/BonusNumber.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.List;
+
+public class BonusNumber {
+    private static final int MIN_NUMBER = 1;
```
로또 최대 최소 값이 BonusNumber, WinningNumbers, LottoGenerator에 중복으로 퍼져있네요
로또 번호를 관리하는 책임은 어느 클래스가 갖고 있는게 좋을까요?

kang-hyungu commented on src/main/java/lotto/domain/BonusNumber.java:
```diff
@@ -0,0 +1,47 @@
+        this.bonus = getValidatedBonusNumber(bonusNumberInput);
+    }
+
+    private int getValidatedBonusNumber(String bonusNumberInput) {
```
메서드는 한 가지 일만 하는게 좋습니다.
메서드명에서 이미 2가지 일을 하는걸로 보이네요
getValidated <= 값을 반환하는 것과 검증 하는 두 가지 일을 하고 있습니다.
두 가지 기능을 분리해보는게 좋겠네요

kang-hyungu commented on src/main/java/lotto/domain/BonusNumber.java:
```diff
@@ -0,0 +1,47 @@
+
+    private int bonus;
+
+    public BonusNumber(String bonusNumberInput) {
```
변수명에 input을 쓸 필요가 있을까요?
입력을 io로만 받진 않을 것 같네요

kang-hyungu commented on src/main/java/lotto/domain/BonusNumber.java:
```diff
@@ -0,0 +1,47 @@
+        return Integer.parseInt(bonusNumberInput);
+    }
+
+    private void checkNoInput(String bonusNumberInput) {
```
noInput는 어색하네요
메서드명에 empty를 쓰면 어떨까요

kang-hyungu commented on src/main/java/lotto/domain/BonusNumber.java:
```diff
@@ -0,0 +1,47 @@
+        }
+    }
+
+    public boolean isIncluded(List<Integer> numbers) {
```
numbers는 BonusNumber 객체에서 관리하는 인스턴스 변수가 아니다 보니 null인지 아닌지 확신할 수가 없겠네요
null 체크가 필요합니다

kang-hyungu replied:
is보단 has가 의미에 맞지 않을까요

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+
+public class Lotto {
+    private static final int LOTTO_NUMBERS_COUNT = 6;
+    private static final int WINNING_NUMBERS_COUNT = 6;
```
당첨번호 갯수를 WinningNumbers에서 관리해야 하지 않을까요?
WinningNumbers에 NUMBER_COUNT_PER_LOTTO가 똑같은 의미인 것 같은데 이 변수도 중복 아닐까요?

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+public class Lotto {
+    private static final int LOTTO_NUMBERS_COUNT = 6;
+    private static final int WINNING_NUMBERS_COUNT = 6;
+    private static final int FIVE_MATCH = 5;
```
이 값은 MatchResult.FIVE_MATCH_WITH_BONUS_BALL와 같은 의미 아닐까요?

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+    private final List<Integer> numbers;
+
+    public Lotto(List<Integer> numbers) {
+        this.numbers = numbers;
```
생성자에서 null 체크를 해주는게 좋습니다

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+        this.numbers = numbers;
+    }
+
+    public MatchResult findMatchResult(WinningNumbers winningNumbers, BonusNumber bonusNumber) {
```
find와 match 둘중 하나만 써도 의미가 통하겠네요

kang-hyungu replied:
MatchResult 결과를 반환하는데 find 또는 match가 맞을까요?
상수를 생성하는거 아닐까요?

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+        return MatchResult.of(sameNumberCount);
+    }
+
+    private int calculateSameNumberCountWith(WinningNumbers winningNumbers) {
```
메서드 이름을 좀더 간결하게 바꿔볼까요?

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+
+    public MatchResult findMatchResult(WinningNumbers winningNumbers, BonusNumber bonusNumber) {
+        int sameNumberCount = calculateSameNumberCountWith(winningNumbers);
+        if (sameNumberCount == FIVE_MATCH && bonusNumber.isIncluded(numbers)) {
```
1. 이 조건문도 메서드로 추출하면 어떨까요?
2. sameNumberCount == FIVE_MATCH를 MatchResult에서 처리하도록 변경해보면 어떨까요?

kang-hyungu commented on src/main/java/lotto/domain/LottoGenerator.java:
```diff
@@ -0,0 +1,50 @@
+
+    private static List<Integer> generateLottoNumbers() {
+        List<Integer> cloneAllLottoNumbers = new ArrayList<>();
+        cloneAllLottoNumbers.addAll(allLottoNumbers);
```
```suggestion
        List<Integer> cloneAllLottoNumbers = new ArrayList<>(allLottoNumbers);
```
좀더 간결하게 써볼까요

kang-hyungu commented on src/main/java/lotto/domain/LottoResults.java:
```diff
@@ -0,0 +1,28 @@
+
+    private final Map<MatchResult, Integer> lottoResults;
+
+    public LottoResults(Map<MatchResult, Integer> matchResults) {
```
```suggestion
    public LottoResults(WinningNumbers winningNumbers, BonusNumber bonusNumber) {
```
이렇게 바꿔보면 어떨까요?
컬렉션을 외부에서 받아오지 말고 LottoResults가 일급 콜렉션이니 외부에서 받아온 값들로 로또 결과를 만들 수 있지 않을까요?

kang-hyungu commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,46 @@
+        this.lottos = lottos;
+    }
+
+    public Map<MatchResult, Integer> createMatchResults(WinningNumbers winningNumbers, BonusNumber bonusNumber) {
```
```suggestion
    public LottoResults createMatchResults(WinningNumbers winningNumbers, BonusNumber bonusNumber) {
```
로또 결과를 반환하게 하면 LottoResults 클래스 외에 다른 코드에서 Map<MatchResult, Integer>를 복잡하게 다룰 일이 줄어들거에요
LottoResults에서만 Map<MatchResult, Integer>를 다루도록 코드를 변경해보아요

kang-hyungu commented on src/main/java/lotto/domain/MatchResult.java:
```diff
@@ -0,0 +1,34 @@
+        return Arrays.stream(values())
+                .filter(v -> matchCount == v.matchCount)
+                .findFirst()
+                .orElse(null);
```
null을 반환하면 NPE 발생 원인이 될 수 있습니다
MatchResult에 NONE 같이 찾는 값이 없을 때 반환할 enum을 추가하면 어떨까요?

kang-hyungu commented on src/main/java/lotto/domain/PurchasePrice.java:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    private static int convertToInt(String purchasePriceInput) {
```
변환과 검증 두가지 일을 한번에 처리하지 말고 분리해보면 어떨까요

kang-hyungu commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.stream.Collectors;
+
+public class WinningNumbers {
+    private static final String WINNING_NUMBERS_DELIMITER = ",";
```
도메인 클래스는 알고 있으면 안되는 값이네요
요구사항이 추가되면 다른 환경에서 입력 받는 방식이 ,가 아니라 다른 걸로 바뀔 수도 있지 않을까요?
view로 옮겨보아요

kang-hyungu commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,96 @@
+
+    private final List<Integer> winningNumbers;
+
+    public WinningNumbers(String winningNumbersInput) {
```
변수명에 input은 없어도 되겠네요

kang-hyungu commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+    private List<Integer> getValidatedWinningNumbers(String winningNumbersInput) {
```
이 메서드명도 마찬가지로 바꾸는게 좋겠어요

kang-hyungu commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,96 @@
+
+    private void checkType(List<String> splittedNumbers) {
+        try {
+            splittedNumbers.stream()
```
forEach를 써도 좋을 것 같아요

kang-hyungu commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private void checkRange(List<Integer> winningNumbers) {
+        int validNumberCount = winningNumbers.stream()
```
count를 써보면 어떨까요
```suggestion
        long validNumberCount = winningNumbers.stream()
                .filter(integer -> MIN_NUMBER <= integer && integer <= MAX_NUMBER)
                .count();
```

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.List;
+import java.util.Set;
+
+public class Lotto {
```
Lotto 테스트도 추가해보면 좋겠어요

kang-hyungu reviewed (changes requested):
안녕하세요 쿨라임! 리뷰어 구구입니다 👋 
미션 구현 잘 하셨습니다.
몇 가지 피드백 남겼으니 확인해보시고 궁금한 점 생기면 코멘트나 dm 남겨주세요

icyMojito committed:
[feat] 구입 금액을 다루는 LottoMoney 클래스 생성

- 구입 금액을 다루었던 PurchasePrice라는 클래스 이름이 추상적이어서 LottoMoney로 변경 후 재설계
- 구입한 로또 개수와 수익률을 구할 수 있음

icyMojito committed:
[test] 구입 금액을 다루는 LottoMoney 테스트

icyMojito committed:
[feat] 로또 숫자 한 개를 다루는 LottoNumber 클래스 생성

- 보너스 숫자는 로또 숫자 한 개의 개념 아래에 속하므로 보너스 숫자를 감싼 BonusNumber 클래스 제거
- 로또 숫자 한 개의 입력값을 검증

icyMojito committed:
[test] 로또 숫자 한 개를 다루는 LottoNumber 클래스 테스트

icyMojito committed:
[feat] 로또 한 줄을 다루는 Lotto 클래스 생성

- 당첨 번호는 로또 한 줄의 개념 아래에 속하므로 당첨 번호를 감싼 WinningNumbers 클래스 제거
- 로또 한 줄을 자동으로 생성 가능
- 입력값을 받아 검증 후 로또 한 줄을 수동으로 생성 가능
- 당첨 결과를 생성 가능

icyMojito committed:
[test] 로또 한 줄을 다루는 Lotto 클래스 테스트

icyMojito committed:
[refactor] 로또들을 다루는 Lottos 클래스의 책임 수정

- 구입 금액만큼 로또 한 줄을 자동 생성하여 보관. 따라서 LottoGenerator 클래스가 필요없어짐
- 로또들의 당첨 결과를 생성 가능

icyMojito committed:
[refactor] 당첨 결과를 관리하는 MatchResult 클래스

- NPE가 발생하지 않도록 해당되는 당첨 결과가 없으면 예외를 던지도록 수정
- 당첨 번호와 일치하는 번호가 5개일 때의 당첨 결과를 MatchResult가 직접 판단하여 알려줄 수 있도록 수정

icyMojito committed:
[feat] 당첨 결과들을 다루는 MatchResults 클래스 생성

- 기존에 당첨 결과들을 다뤘던 LottoResults 클래스 이름은 추상적이어서 삭제
- 로또들의 당첨 결과를 받아 통계를 만들어 관리
- 당첨 결과 통계로 총 수익금 계산 가능

icyMojito committed:
[test] 당첨 결과들을 다루는 MatchResults 클래스 테스트

icyMojito committed:
[feat] 공통된 입력값 검증을 책임지는 NumberValidator 클래스 생성

LottoMoney 클래스와 LottoNumber 클래스에서 중복되는 입력값 검증을 수행

icyMojito committed:
[refactor] InputView에서 구분자를 다루도록 수정

구분자는 InputView의 관심사이기 때문에 InputView에서 구분자 정보를 처리하도록 수정

icyMojito committed:
[refactor] OutputView 메서드명 수정

어떤 정보를 찍는지 정확하게 표현을 위해 메서드명 수정

icyMojito committed:
[refactor] LottoApplication 클래스 정리

재설계한 클래스에 따라 코드를 정리

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.*;
+
+public class Lotto {
+    public static final int LOTTO_NUMBERS_COUNT = 6;
```
접근제어자를 public으로 공개할 필요는 없어보여요

icyMojito replied:
새로 만든 LottoFactory라는 클래스에서 LOTTO_NUMBERS_COUNT가 필요하게 되어 
부득이하게 접근제어자를 public으로 두게 되었습니다~

kang-hyungu replied:
앗 그러네요 default로 되는줄 알았는데 public으로 열어놔야겠네요 😅

kang-hyungu commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,78 @@
+
+    private static List<LottoNumber> createLottoNumbers() {
+        Set<LottoNumber> lottoNumbers = new HashSet<>();
+        while (lottoNumbers.size() != LOTTO_NUMBERS_COUNT) {
```
Lotto 객체를 생성할 때마다 랜덤 객체를 사용해서 겹치지 않는 6개 숫자가 나올때까지 반복해야겠네요
성능 개선을 위해 미리 1부터 45까지 LottoNumber를 map으로 만들어놓고 Collections.shuffle 메서드를 활용해서 6자리 숫자를 복사해오는 방식으로 바꿔보면 어떨까요?
일종의 LottoNumber 캐시를 만들어보아요

kang-hyungu commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,29 @@
+public class Lottos {
+    private final List<Lotto> lottos;
+
+    public Lottos(LottoMoney money) {
```
파라미터로 객체가 넘어올 때 null이면 어떻게 처리할지 꼭 챙겨야합니다!

kang-hyungu commented on src/main/java/lotto/domain/MatchResult.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    static MatchResult of(int matchCount, boolean containsBonusNumber) {
+        if (matchCount != 5) {
```
매직넘버!
5 대신 FIVE_MATCH.matchCount를 쓰면 어떨까요?

kang-hyungu reviewed (changes requested):
피드백 반영 잘 하셨어요 👍 
코드가 전반적으로 더 깔끔해졌네요
한 가지 더 챙기면 좋을 부분이 있어서 피드백 남겼어요
확인해보시고 궁금한 점 생기면 코멘트나 dm 남겨주세요!

icyMojito committed:
[feat] 로또 한 줄을 자동으로 뽑아주는 LottoFactory 클래스 생성

성능 개선을 위해 미리 만들어 둔 로또 숫자들을 가지고 로또 한 줄을 만들어주는 클래스 생성

icyMojito committed:
[refactor] 로또 한 줄을 자동 생성하는 방법의 변경에 따라 Lotto 클래스 생성자 수정

성능 개선을 위해 랜덤 객체를 사용하지 않는 방법을 이용하여 로또 한 줄을 생성할 수 있도록 생성자 수정

icyMojito committed:
[refactor] LottoNumber 클래스에서 로또 숫자 한 개를 자동생성하지 않도록 수정

로또 숫자 한 개를 자동 생성하기 위해 랜덤 객체를 사용하면 성능이 낮아지므로 해당 방법은 제거함

icyMojito committed:
[refactor] Lottos의 생성자 매개변수 Null 체크 기능 추가

icyMojito committed:
[refactor] MatchResult 클래스의 매직넘버를 클래스 변수화 처리

kang-hyungu reviewed (changes requested):
피드백 반영 잘 하셨어요 👍 
그런데 수정하면서 놓친 부분 같은데 컴파일 에러가 발생하는 클래스가 보이네요
이부분만 수정해서 다시 요청주시면 머지 처리할게요

icyMojito committed:
[refactor] Lotto 클래스 컴파일 에러 해결

kang-hyungu reviewed (approved):
피드백 반영하느라 수고하셨어요!
pr은 머지 처리할게요
